### DISCUSSION TITLE
How to convert (N,3) numpy array into (N) warp vec3?

### DISCUSSION THREAD

**Discussion by cadop** [NVIDIA/warp#359] [Q&A]
I am trying to convert a numpy array of integers into a warp array, where the second dim of numpy is captured by the warp dtype. 

```
    faces = mesh.indices.reshape((-1, 3))
    faces = wp.array(faces, shape=(N), dtype=wp.vec3i, device=device)
```

The kernel signature is:
```
@wp.kernel
def _matrix(points: wp.array(dtype=wp.vec3),
            faces: wp.array(dtype=wp.vec3i),
            mesh: wp.uint64):
     f = wp.tid()
     face = faces[f]
```

This was seemingly working when I was in `device='cpu'` , however changing device to cuda results in:
 `RuntimeError: Error launching kernel '_matrix', argument 'faces' expects an array with dtype=<class 'warp.types.vec3i'> but passed array has dtype=<class 'warp.types.int32'>.`

**Comment by cadop** [comment]
As a sanity check, this does work:
```
    faces = mesh.indices.reshape((-1, 3))
    faces = wp.array(faces, shape=(N,3), dtype=wp.int32, device=device)
```

```
@wp.kernel
def _matrix(points: wp.array(dtype=wp.vec3),
            faces: wp.array(dtype=wp.int32, ndim=2),
            mesh: wp.uint64):
     f = wp.tid()
     face = faces[f]
```

**Comment by christophercrouzet** [comment]
Hi @cadop,

Would you be able to share which version of Warp you've been using and sharing a runnable snippet showcasing the issue?

I didn't manage to repro this issue on my end using the following snippet:

```py
import numpy as np
import warp as wp


POINTS = (
    ( 0.853553, -0.146446,  0.0     ), ( 0.146446, -0.853553,  0.0     ),
    ( 0.353553,  0.353553,  0.707106), (-0.353553, -0.353553,  0.707106),
    (-0.353553, -0.353553, -0.707106), ( 0.353553,  0.353553, -0.707106),
    (-0.853553,  0.146446,  0.0     ), (-0.146446,  0.853553,  0.0     ),
)

FACE_VERTEX_INDICES = (
    0, 3, 1, 0, 2, 3, 4, 7, 5, 4, 6, 7, 6, 2, 7, 6, 3, 2,
    5, 1, 4, 5, 0, 1, 5, 2, 0, 5, 7, 2, 1, 6, 4, 1, 3, 6,
)


@wp.kernel
def _matrix(
    points: wp.array(dtype=wp.vec3),
    faces: wp.array(dtype=wp.vec3i),
):
    f = wp.tid()
    face = faces[f]
    wp.printf("face=(%d, %d, %d)\n", face[0], face[1], face[2])


def run():
    points = wp.array(POINTS, dtype=wp.vec3)

    faces = np.array(FACE_VERTEX_INDICES).reshape((-1, 3))
    faces = wp.array(faces, shape=faces.shape[0], dtype=wp.vec3i)

    wp.launch(_matrix, dim=faces.shape, inputs=(points, faces))
    wp.synchronize()


if __name__ == "__main__":
    run()
```

Note that these 2 lines:
```py
faces = np.array(FACE_VERTEX_INDICES).reshape((-1, 3))
faces = wp.array(faces, shape=faces.shape[0], dtype=wp.vec3i)
```

Could be replaced by single call:
```py
faces = wp.array(FACE_VERTEX_INDICES, dtype=wp.vec3i)
```

**Reply by cadop** [reply]
Hi Christopher, sorry for the missing context. I opened a discussion instead of an issue because I assumed I just didn't do something right rather than saying it was a bug. 

I think the issue I have is because I didn't realize I wasn't using numpy, my array was converted to warp at some point. I was trying to convert a warp array of dtype `int` to `vec3i`.  I guess my question would be how to do that, or if `view` works similar to numpy and can be passed to kernels (I wasn't able to view it though).

```
import numpy as np
import warp as wp

device = wp.get_device("cuda:1")

POINTS = (
    ( 0.853553, -0.146446,  0.0     ), ( 0.146446, -0.853553,  0.0     ),
    ( 0.353553,  0.353553,  0.707106), (-0.353553, -0.353553,  0.707106),
    (-0.353553, -0.353553, -0.707106), ( 0.353553,  0.353553, -0.707106),
    (-0.853553,  0.146446,  0.0     ), (-0.146446,  0.853553,  0.0     ),
)

FACE_VERTEX_INDICES = (
    0, 3, 1, 0, 2, 3, 4, 7, 5, 4, 6, 7, 6, 2, 7, 6, 3, 2,
    5, 1, 4, 5, 0, 1, 5, 2, 0, 5, 7, 2, 1, 6, 4, 1, 3, 6,
)


@wp.kernel
def _matrix(
    points: wp.array(dtype=wp.vec3),
    faces: wp.array(dtype=wp.vec3i),
):
    f = wp.tid()
    face = faces[f]
    wp.printf("face=(%d, %d, %d)\n", face[0], face[1], face[2])


def run():
    points = wp.array(POINTS, dtype=wp.vec3)

    faces = wp.array(FACE_VERTEX_INDICES).reshape((-1, 3))
    faces = wp.array(faces, shape=faces.shape[0], dtype=wp.vec3i)

    wp.launch(_matrix, dim=faces.shape, inputs=(points, faces))
    wp.synchronize()


if __name__ == "__main__":
    run()
```

**Reply by christophercrouzet** [reply]
Hi @cadop, thanks for providing this snippet, I've been able to repro the issue and your approach should be valid, so this does look like a bug. Would you be able to open an issue so that we can track it appropriately?

Regarding the `view()` method, if you look at its implementation, it's basically a shortcut to calling `wp.array()` while ensuring that no copy is being made, so yes, these array views can be passed to kernels. However, as you've noticed, it's failing here since it expects the new `dtype` to have the same width as the old one, so directly calling `wp.array()` as you've done it brings more flexibility here.

**Reply by cadop** [reply]
This is resolved right? 
https://github.com/NVIDIA/warp/issues/363#issuecomment-2599382262

**Reply by christophercrouzet** [reply]
Yes, it should be!
